Assert evaluating expression after invalid using in Interactive Window

```
> from x in args select x + 1
(1,11): error CS0103: The name 'args' does not exist in the current context
> var args = new[] { 1, 2, 3 };
> from x in args select x + 1
(1,11): error CS1935: Could not find an implementation of the query pattern for source type 'int[]'.  'Select' not found.  Are you missing a reference to 'System.Core.dll' or a using directive for 'System.Linq'?
> using System.Linq;
> from x in args select x + 1
Unexpected error:
System.InvalidOperationException: Previous submission has errors.
   at Microsoft.CodeAnalysis.Compilation.ValidateSubmissionParameters(Compilation previousSubmission, Type returnType, Type& hostObjectType) in F:\Enlistments\roslyn\src\Compilers\Core\Portable\Compilation\Compilation.cs:line 148
   at Microsoft.CodeAnalysis.CSharp.CSharpCompilation.Create(String assemblyName, CSharpCompilationOptions options, IEnumerable`1 syntaxTrees, IEnumerable`1 references, CSharpCompilation previousSubmission, Type returnType, Type hostObjectType, Boolean isSubmission) in F:\Enlistments\roslyn\src\Compilers\CSharp\Portable\Compilation\CSharpCompilation.cs:line 247
   at Microsoft.CodeAnalysis.CSharp.CSharpCompilation.CreateSubmission(String assemblyName, SyntaxTree syntaxTree, IEnumerable`1 references, CSharpCompilationOptions options, Compilation previousSubmission, Type returnType, Type hostObjectType) in F:\Enlistments\roslyn\src\Compilers\CSharp\Portable\Compilation\CSharpCompilation.cs:line 222
   at Microsoft.CodeAnalysis.Scripting.CSharp.CSharpScript.CreateCompilation() in F:\Enlistments\roslyn\src\Scripting\CSharp\CSharpScript.cs:line 155
   at Microsoft.CodeAnalysis.Scripting.Script.GetCompilation() in F:\Enlistments\roslyn\src\Scripting\Core\Script.cs:line 350
   at Microsoft.CodeAnalysis.Scripting.Script.GetExecutor(CancellationToken cancellationToken) in F:\Enlistments\roslyn\src\Scripting\Core\Script.cs:line 408
   at Microsoft.CodeAnalysis.Scripting.Script.Build() in F:\Enlistments\roslyn\src\Scripting\Core\Script.cs:line 363
   at Microsoft.CodeAnalysis.Interactive.InteractiveHost.Service.Compile(String text, String path) in F:\Enlistments\roslyn\src\Interactive\Features\Interactive\Core\InteractiveHost.Service.cs:line 655
   at Microsoft.CodeAnalysis.Interactive.InteractiveHost.Service.Execute(String text) in F:\Enlistments\roslyn\src\Interactive\Features\Interactive\Core\InteractiveHost.Service.cs:line 736
   at Microsoft.CodeAnalysis.Interactive.InteractiveHost.Service.ExecuteAsync(RemoteAsyncOperation`1 operation, String text) in F:\Enlistments\roslyn\src\Interactive\Features\Interactive\Core\InteractiveHost.Service.cs:line 381
```

--- a/src/Compilers/CSharp/Portable/Compilation/CSharpCompilation.cs
+++ b/src/Compilers/CSharp/Portable/Compilation/CSharpCompilation.cs
@@ -2250,12 +2250,7 @@ namespace Microsoft.CodeAnalysis.CSharp
             DiagnosticBag diagnostics,
             CancellationToken cancellationToken)
         {
-            // Do not waste a slot in the submission chain for submissions that contain no executable code
-            // (they may only contain #r directives, usings, etc.)
-            if (IsSubmission && !HasCodeToEmit())
-            {
-                return null;
-            }
+            Debug.Assert(!IsSubmission || HasCodeToEmit());
 
             string runtimeMDVersion = GetRuntimeMetadataVersion(emitOptions, diagnostics);
             if (runtimeMDVersion == null)

--- a/src/Compilers/Core/Portable/Compilation/Compilation.cs
+++ b/src/Compilers/Core/Portable/Compilation/Compilation.cs
@@ -1579,6 +1579,15 @@ namespace Microsoft.CodeAnalysis
                 return ToEmitResultAndFree(diagnostics, success: false, entryPointOpt: null);
             }
 
+            // Do not waste a slot in the submission chain for submissions that contain no executable code
+            // (they may only contain #r directives, usings, etc.)
+            if (IsSubmission && !HasCodeToEmit())
+            {
+                // Still report diagnostics since downstream submissions will assume there are no errors.
+                diagnostics.AddRange(this.GetDiagnostics());
+                return ToEmitResultAndFree(diagnostics, success: false, entryPointOpt: null);
+            }
+
             var moduleBeingBuilt = this.CreateModuleBuilder(
                 options,
                 manifestResources,

--- a/src/Compilers/VisualBasic/Portable/Compilation/VisualBasicCompilation.vb
+++ b/src/Compilers/VisualBasic/Portable/Compilation/VisualBasicCompilation.vb
@@ -2147,12 +2147,7 @@ Namespace Microsoft.CodeAnalysis.VisualBasic
             cancellationToken As CancellationToken) As CommonPEModuleBuilder
 
             Debug.Assert(diagnostics.IsEmptyWithoutResolution) ' True, but not required.
-
-            ' Do not waste a slot in the submission chain for submissions that contain no executable code
-            ' (they may only contain #r directives, usings, etc.)
-            If IsSubmission AndAlso Not HasCodeToEmit() Then
-                Return Nothing
-            End If
+            Debug.Assert(Not IsSubmission OrElse HasCodeToEmit())
 
             ' Get the runtime metadata version from the cor library. If this fails we have no reasonable value to give.
             Dim runtimeMetadataVersion = GetRuntimeMetadataVersion()

--- a/src/Scripting/CSharpTest/InteractiveSessionTests.cs
+++ b/src/Scripting/CSharpTest/InteractiveSessionTests.cs
@@ -1996,6 +1996,33 @@ class D
             Assert.Throws<ArgumentException>(() => CSharpCompilation.CreateSubmission("a", options: TestOptions.ReleaseDll.WithDelaySign(false)));
         }
 
+        [WorkItem(3795, "https://github.com/dotnet/roslyn/issues/3795")]
+        [Fact]
+        public void ErrorInUsing()
+        {
+            var submission = CSharpCompilation.CreateSubmission("sub1", Parse("using Unknown;", options: TestOptions.Script), new[] { MscorlibRef });
+
+            var expectedDiagnostics = new[]
+            {
+                    // (1,7): error CS0246: The type or namespace name 'Unknown' could not be found (are you missing a using directive or an assembly reference?)
+                    // using Unknown;
+                    Diagnostic(ErrorCode.ERR_SingleTypeNameNotFound, "Unknown").WithArguments("Unknown").WithLocation(1, 7),
+                    // (1,1): hidden CS8019: Unnecessary using directive.
+                    // using Unknown;
+                    Diagnostic(ErrorCode.HDN_UnusedUsingDirective, "using Unknown;").WithLocation(1, 1),
+            };
+
+            // Emit produces the same diagnostics as GetDiagnostics (below).
+            using (var stream = new MemoryStream())
+            {
+                var emitResult = submission.Emit(stream);
+                Assert.False(emitResult.Success);
+                emitResult.Diagnostics.Verify(expectedDiagnostics);
+            }
+
+            submission.GetDiagnostics().Verify(expectedDiagnostics);
+        }
+
         private CSharpCompilation CreateSubmission(string code, CSharpParseOptions options, int expectedErrorCount = 0)
         {
             var submission = CSharpCompilation.CreateSubmission("sub",

--- a/src/Scripting/VisualBasicTest/InteractiveSessionTests.vb
+++ b/src/Scripting/VisualBasicTest/InteractiveSessionTests.vb
@@ -402,6 +402,27 @@ End Function")
             'Assert.False(symbols.Any(Function(s) s.Name = "Roslyn"))
         End Sub
 
+        <WorkItem(3795, "https:'github.com/dotnet/roslyn/issues/3795")>
+        <Fact>
+        Public Sub ErrorInUsing()
+            Dim submission = VisualBasicCompilation.CreateSubmission("sub1", Parse("Imports Unknown", options:=TestOptions.Script), {MscorlibRef})
+
+            Dim expectedErrors = <errors><![CDATA[
+BC40056: Namespace or type specified in the Imports 'Unknown' doesn't contain any public member or cannot be found. Make sure the namespace or the type is defined and contains at least one public member. Make sure the imported element name doesn't use any aliases.
+Imports Unknown
+        ~~~~~~~
+]]></errors>
+
+            ' Emit produces the same diagnostics as GetDiagnostics (below).
+            Using stream As New MemoryStream()
+                Dim emitResult = submission.Emit(stream)
+                Assert.False(emitResult.Success)
+                emitResult.Diagnostics.AssertTheseDiagnostics(expectedErrors)
+            End Using
+
+            submission.GetDiagnostics().AssertTheseDiagnostics(expectedErrors)
+        End Sub
+
 #End Region
 
 #Region "Anonymous types"

